chainstate/src/detail/transaction_verifier/mod.rs

```diff
@@ -201,60 +215,98 @@ impl<'a, S: BlockchainStorageRead> TransactionVerifier<'a, S> {
                     .ok_or(ConnectTransactionError::MissingOutputOrSpent)?;
                 entry.insert(CachedInputsOperation::Read(tx_index));
             }
-        }
+        };
         Ok(())
     }
 
-    fn check_transferred_amounts_and_get_fee(
-        &self,
-        tx: &Transaction,
-    ) -> Result<Fee, ConnectTransactionError> {
-        let inputs = tx.inputs();
-        let outputs = tx.outputs();
-
-        let inputs_total = self.calculate_coins_total_inputs(inputs)?;
-        let outputs_total = Self::calculate_coins_total_outputs(outputs)?;
-
-        if outputs_total > inputs_total {
-            return Err(ConnectTransactionError::AttemptToPrintMoney(
-                inputs_total,
-                outputs_total,
-            ));
-        }
+    fn calculate_total_outputs(
+        outputs: &[TxOutput],
+        include_issuance: Option<&Transaction>,
+    ) -> Result<BTreeMap<CoinOrTokenId, Amount>, ConnectTransactionError> {
+        let mut total_amounts = BTreeMap::new();
 
-        let paid_fee = inputs_total - outputs_total;
-        paid_fee.map(Fee).ok_or(ConnectTransactionError::TxFeeTotalCalcFailed(
-            inputs_total,
-            outputs_total,
-        ))
+        outputs
+            .iter()
+            .map(|output| get_output_token_id_and_amount(output.value(), include_issuance))
+            .try_for_each(|item| {
+                item?.map(|(coin_or_token, amount)| {
+                    insert_or_increase(&mut total_amounts, coin_or_token, *amount)
+                });
+                Ok::<(), TokensError>(())
+            })?;
+        Ok(total_amounts)
     }
 
-    fn calculate_coins_total_inputs(
+    fn calculate_total_inputs(
         &self,
         inputs: &[TxInput],
-    ) -> Result<Amount, ConnectTransactionError> {
-        inputs.iter().try_fold(Amount::from_atoms(0), |total, input| {
-            let amount = self
+    ) -> Result<BTreeMap<CoinOrTokenId, Amount>, ConnectTransactionError> {
+        let mut result = BTreeMap::new();
+
+        for input in inputs {
+            let utxo = self
                 .utxo_cache
                 .utxo(input.outpoint())
-                .ok_or(ConnectTransactionError::MissingOutputOrSpent)
-                .map(|utxo| match utxo.output().value() {
-                    OutputValue::Coin(amount) => *amount,
-                })?;
-            (total + amount).ok_or(ConnectTransactionError::OutputAdditionError)
-        })
+                .ok_or(ConnectTransactionError::MissingOutputOrSpent)?;
+
+            match input.outpoint().tx_id() {
+                OutPointSourceId::Transaction(issuance_tx_id) => {
```

Comment:
So every OutPointSourceId's parameter is an issuance_tx_id? If I understand correctly, this may or may not be an issuance transaction, we don't know yet. It's just an outpoint tx id.

Reply:
Sam, you fixed this